controllers/cluster_controller_phases_test.go

```diff
@@ -370,3 +370,178 @@ func TestClusterReconciler_reconcilePhase(t *testing.T) {
 		})
 	}
 }
+
+func TestClusterReconcilePhases_reconcileFailureDomains(t *testing.T) {
+	cluster := &clusterv1.Cluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test-cluster",
+			Namespace: "test-namespace",
+		},
+		Status: clusterv1.ClusterStatus{
+			InfrastructureReady: true,
+		},
+		Spec: clusterv1.ClusterSpec{
+			ControlPlaneEndpoint: clusterv1.APIEndpoint{
+				Host: "1.2.3.4",
+				Port: 8443,
+			},
+			InfrastructureRef: &corev1.ObjectReference{
+				APIVersion: "infrastructure.cluster.x-k8s.io/v1beta1",
+				Kind:       "GenericInfrastructureMachine",
+				Name:       "test",
+			},
+		},
+	}
+
+	newFailureDomain := clusterv1.FailureDomains{
+		"newdomain": clusterv1.FailureDomainSpec{
+			ControlPlane: false,
+			Attributes: map[string]string{
+				"attribute1": "value1",
+			},
+		},
+	}
+
+	newFailureDomainUpdated := clusterv1.FailureDomains{
+		"newdomain": clusterv1.FailureDomainSpec{
+			ControlPlane: false,
+			Attributes: map[string]string{
+				"attribute2": "value2",
+			},
+		},
+	}
+
+	clusterWithNewFailureDomainUpdated := cluster.DeepCopy()
+	clusterWithNewFailureDomainUpdated.Status.FailureDomains = newFailureDomainUpdated
+
+	oldFailureDomain := clusterv1.FailureDomains{
+		"olddomain": clusterv1.FailureDomainSpec{
+			ControlPlane: false,
+			Attributes: map[string]string{
+				"attribute1": "value1",
+			},
+		},
+	}
+
+	clusterWithOldFailureDomain := cluster.DeepCopy()
+	clusterWithOldFailureDomain.Status.FailureDomains = oldFailureDomain
+
+	tests := []struct {
+		name                 string
+		cluster              *clusterv1.Cluster
+		infraRef             map[string]interface{}
+		expectFailureDomains clusterv1.FailureDomains
+	}{
+		{
+			name:    "expect no failure domain if infrastructure ref is nil",
+			cluster: &clusterv1.Cluster{ObjectMeta: metav1.ObjectMeta{Name: "test-cluster", Namespace: "test-namespace"}},
+		},
+		{
+			name:    "expect no failure domain if infra config does not have failure domain",
+			cluster: cluster,
+			infraRef: map[string]interface{}{
```

Comment:
as a follow up consider to have a helper for building these unstructured infraRefs.

Reply:
Ack 

@enxebre  initial lint failures were due to 

https://github.com/kubernetes-sigs/cluster-api/pull/5656

fixed in this PR as well

Reply:
Removed the lint fix as pr is merged